pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,296 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/validation"
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/util/shufflesharding"
+	apivalidation "k8s.io/kubernetes/pkg/apis/core/validation"
+	"k8s.io/kubernetes/pkg/apis/flowcontrol"
+)
+
+// ValidateFlowSchemaName validates name for flow-schema.
+var ValidateFlowSchemaName = apimachineryvalidation.NameIsDNSSubdomain
+
+// ValidatePriorityLevelConfigurationName validates name for priority-level-configuration.
+var ValidatePriorityLevelConfigurationName = apimachineryvalidation.NameIsDNSSubdomain
+
+var supportedDistinguisherMethods = sets.NewString(
+	string(flowcontrol.FlowDistinguisherMethodByNamespaceType),
+	string(flowcontrol.FlowDistinguisherMethodByUserType),
+)
+
+var supportedVerbs = sets.NewString(
+	flowcontrol.VerbAll,
+	"get",
+	"list",
+	"watch",
+	"create",
+	"update",
+	"delete",
+	"deletecollection",
+	"patch",
+)
+
+var supportedSubjectKinds = sets.NewString(
+	string(flowcontrol.SubjectKindServiceAccount),
+	string(flowcontrol.SubjectKindGroup),
+	string(flowcontrol.SubjectKindUser),
+)
+
+var supportedQueuingType = sets.NewString(
+	string(flowcontrol.PriorityLevelQueuingTypeQueueing),
+	string(flowcontrol.PriorityLevelQueuingTypeExempt),
+)
+
+// ValidateFlowSchema validates the content of flow-schema
+func ValidateFlowSchema(fs *flowcontrol.FlowSchema) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&fs.ObjectMeta, false, ValidateFlowSchemaName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateFlowSchemaSpec(&fs.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateFlowSchemaStatus(&fs.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidateFlowSchemaSpec validates the content of flow-schema's spec
+func ValidateFlowSchemaSpec(spec *flowcontrol.FlowSchemaSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if spec.MatchingPrecedence < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("matchingPrecedence"), spec.MatchingPrecedence, "must be positive value"))
```

Comment:
A value of `0` is never going to appear here, the defaulting code ensures that.

Reply:
yes, the defaulting will make sure 0 wont happen, are we validating that anyway? @lavalamp

Reply:
the same question applies all the suggestions on `<= 0`, they all have a default now

Reply:
We have decided to go with checking whether the value is positive and saying "must be positive" in the error message, because (1) this removes the "long-range dependency" on defaulting code and (2) the error message is consistent with the check.